#159987631 Users should be able to comment on articles

CEOehis opened the pull request:
#### What does this PR do?
Implement ability for users to comment on articles

#### Description of Task to be completed?
Only authenticated users can to comments on articles.
Successive comments are displayed as threaded Timestamped comments.
Render the list of comments belonging to the article.

#### How should this be manually tested?
- Log in to the app
- Navigate to `localhost:3000/aritcle/:articleSlug`
- View and create comments


#### What are the relevant pivotal tracker stories?
[#159987631](https://www.pivotaltracker.com/story/show/159987631)
#### Screenshots (if appropriate)
![localhost_3001_article_wildlife-conservation](https://user-images.githubusercontent.com/25967737/46856818-63d41c80-cdff-11e8-9d7d-5e400214e1a4.png)

![localhost_3001_article_wildlife-conservation 1](https://user-images.githubusercontent.com/25967737/46856772-430bc700-cdff-11e8-93c1-29c30779c8f0.png)


![localhost_3001_article_wildlife-conservation 2](https://user-images.githubusercontent.com/25967737/46856778-499a3e80-cdff-11e8-8a37-ff498cfd8ae0.png)

CEOehis committed:
feat(comment): Add comment box component

Create comment input component to handle creation of comments

CEOehis committed:
feat(comment): separate comments into components

CEOehis committed:
fix: Set api url in .env

CEOehis committed:
feat(comments): fetch article comments from api

Get article slug from window.location and fetch comments for the article.
Show spinning icon while fetching comments

CEOehis committed:
feat(comments): handle fetch article comments

Render dynamic comments
Include relevant tests

CEOehis committed:
feat(comments): implement ability to comment on articles

- Display successive comments as timestamped threads
- Display comment button when user begins to type a new comment
- Render new comment as top most commnet
- Include relevant tests
[Delivers #159987631]

lauragift21 commented on src/components/ArticleComment/ArticleComment.js:
```diff
@@ -1,45 +1,56 @@
+})
 
-export default ArticleComment;
+export default connect(mapStateToProps, mapDispatchToProps, null, { pure: false })(ArticleComment);
```
Why are you passing null to the connection here?  This doesn't work without the null statement `connect(mapStateToProps, mapDispatchToProps)(ArticleComment)`?

CEOehis replied:
I needed to set it to be an impure component to allow rendering because for some reason it was not re-rendering when `props` was updated. I referrenced https://stackoverflow.com/questions/43520498/react-router-private-routes-redirect-not-working

lauragift21 commented on src/components/Comment/Comments.js:
```diff
@@ -0,0 +1,21 @@
+import Comment from './Comment';
+
+const Comments = ({ comments, currentArticleTitle }) => (
+  <React.Fragment>
```
Since this was as a result of not importing Fragment like so , you can do this `import React, { Fragment } from 'react'` so you can then use just `<Fragment></Fragment>` to refactor the code in line 6.

CEOehis replied:
I agree. I'll fix and push.

lauragift21 commented on src/index.js:
```diff
@@ -9,6 +9,7 @@ import userIsLoggedIn from './utils/userIsLoggedIn';
 import { logOutUser, setLoggedInUser } from './actions/auth';
 
 const store = configureStore();
+console.log('apiUrl', process.env.REACT_APP_API);
```
You might want to remove this console.log statement

CEOehis replied:
I put this console.log statement so that when the app is running we can know what api it is making ajax requests to.

lauragift21 replied:
Alright, That makes sense.

lauragift21 reviewed (changes requested):
LGTM 👍

AdmiralOnu commented on src/store/initialState.js:
```diff
@@ -10,6 +10,13 @@ const initialState = {
+    fetchingCommentsError: '',
+    currentArticleComments: [],
+    postingNewComment: false,
+    newCommentError: ''
```
couldn't `postingNewComment` and `fetchingArticleComments` just be `loading`, and `newCommentError` and `fetchingCommentsError` as `error`.  This will also lead to reducing the `createCommentRequest()`, `getArticleCommentsLoading()`, `getArticleCommentsError()` `createCommentError()` functions from actions to something like `setLoading()` and `setError()`

CEOehis replied:
Yeah. I agree. 

However, I did it this way because I needed to handle error states for when the fetching comments, errors out and also for when posting a new comment, errors out. This is a bit more explicit as well.

johnkingzy reviewed (changes requested):
Nice one! 
I will suggest you refactor your store object to something like 
```
comments: {
  articleComments: {
    data: [],
    error: '',
    loading: false
  },
  newComment: {
    data: [],
    error: '',
    loading: false,
  }
 }
```
So when it's fetching all articles, only `articleComments` will be updated, then `newComment` will be use while adding a new comment. 👍

CEOehis committed:
nitpicks: separate articleComments and newComment in store

- UX improvements - show errors when fetching comments fail
- Show errors when posting comments fail

johnkingzy commented on src/reducers/index.js:
```diff
@@ -2,11 +2,14 @@ import { combineReducers } from 'redux';
   auth,
   signinuser,
   signup,
+  articleComments,
```
The initial state has the `articleComments` and `newComment` nested inside the `comments` object. But the rootReducer is not implementing the same structure. Please fix this.

CEOehis replied:
Thanks. I've fixed it.

CEOehis committed:
nitpick - move comments related state to comments state
